integration/common/openlineage/common/provider/dbt/structured_logs.py

```diff
@@ -0,0 +1,515 @@
+            stacktrace = event["data"]["exc_info"]
+            run_facets["errorMessage"] = error_message_run.ErrorMessageRunFacet(
+                message=error_message, programmingLanguage="python", stackTrace=stacktrace
+            )
+        elif dbt_node_type == "SQLQueryStatus":
+            run_state = RunState.COMPLETE
+
+        return generate_run_event(
+            event_type=run_state,
+            event_time=event_time,
+            run_id=sql_ol_run_event.run.runId,
+            run_facets=run_facets,
+            job_name=sql_ol_run_event.job.name,
+            job_namespace=sql_ol_run_event.job.namespace,
+            job_facets=sql_ol_run_event.job.facets,
+        )
+
+    def _parse_command_completed_event(self, event) -> RunEvent:
+        success = event["data"]["success"]
+        event_time = get_event_timestamp(event["data"]["completed_at"])
+        run_facets = self.dbt_version_facet()
+        if success:
+            run_state = RunState.COMPLETE
+        else:
+            run_state = RunState.FAIL
+            error_message = event["info"]["msg"]
+            run_facets["errorMessage"] = error_message_run.ErrorMessageRunFacet(
+                message=error_message, programmingLanguage="python"
+            )
+
+        parent_run_metadata = get_parent_run_metadata()
+
+        if parent_run_metadata:
+            run_facets["parent"] = parent_run_metadata
+
+        return generate_run_event(
+            event_type=run_state,
+            event_time=event_time,
+            run_id=self.dbt_run_metadata.run_id,
+            job_name=self.dbt_run_metadata.job_name,
+            job_namespace=self.dbt_run_metadata.job_namespace,
+            run_facets=run_facets,
+        )
+
+    def _setup_dbt_run_metadata(self, start_event):
+        """
+        The dbt command defines the parent run metadata of
+        all subsequent OL events (NodeStart, NodeFinish ...)
+        """
+        self.dbt_run_metadata = ParentRunMetadata(
+            run_id=start_event.run.runId,
+            job_name=start_event.job.name,
+            job_namespace=start_event.job.namespace,
+        )
+
+    def dbt_version_facet(self) -> DbtVersionRunFacet:
+        return {"dbt_version": DbtVersionRunFacet(version=self.dbt_version)}
+
+    def _get_sql_query_id(self, node_id) -> str:
+        """
+        Not all adapters have the sql id defined in their dbt event.
+        A node is executed by a single thread which means that sql queries of a single node are executed
+        sequentially and their status is also reported sequentially.
+        This function gives us a surrogate query id. It's auto-incremented
+        """
+        if node_id not in self.node_id_to_sql_query_id:
+            self.node_id_to_sql_query_id[node_id] = 1
+
+        query_id = self.node_id_to_sql_query_id[node_id]
+        self.node_id_to_sql_query_id[node_id] += 1
+        return query_id
+
+    def _get_sql_job_name(self, event) -> str:
+        """
```

Comment:
>Can there be more than one SQL events per regular node, and in which case? 

Yes! There can be multiple SQL queries per single Node. Those queries correspond to the materialization used. Example in [Snowflake](https://github.com/dbt-labs/dbt-snowflake/blob/main/dbt/include/snowflake/macros/materializations/incremental.sql#L92C14-L92C23). There are queries for (temp table creation, pre and post hooks and more).

>I'd rather recommend storing this data locally and attaching the SqlJobFacet on COMPLETE event, rather than sending separate events with just the SQL

If we do that we wouldn't get realtime observability on what the node is executing.
The dbt materializations can be customized by users and they can run non-trivial slow SQL in them. We would wait until the execution of the last SQL query to report progress.

>it would make OL consumers lifes harder if they wanted to coalesce that information

It's a handful of SQL queries (< 100 in most cases). I think unlocking real time execution progress is worth the trouble of some extra OL events.

in addition, consumer that want to coalesce information would still do the same for node executions (node lifecycle events plus scheduler events).

Reply:
> It's a handful of SQL queries (< 100 in most cases).

Ah, it's for incremental materialization. Still, that sounds like the "SQL template" is the same, just that it inserts/processes different values, right? They are not exactly "different queries" but different executions of the same template. I might be wrong but I feel then we'd need to talk about some concrete example.

Regarding custom materializations that execute wildly different SQLs, I'm not sure - it sounds like that case makes processing more complex for the most common, simple case, without incremental materialization. I think it's valid use case, but it would be best if handling it would not cause worse experience for 99% of use cases - which IMO would be regular table/ephemeral/materialized_view materializations.

So, maybe we should do
 - if we're using non-incremental materializations, we should attach the sql execution data into regular `START`/`COMPLETE` events
 - if we're using regular incremental materialization, we should intermittently emit `RUNNING` events for the same run - with the same run_id
 - if we have some custom materialization... we should maybe have some custom dbt facet? I feel _in any case_ dbt tries to keep to a model where one dbt node == one table, and single run for that IMO fits the best

Reply:
>Ah, it's for incremental materialization. Still, that sounds like the "SQL template" is the same, just that it inserts/processes different values, right? They are not exactly "different queries" but different executions of the same template. I might be wrong but I feel then we'd need to talk about some concrete example.

Sure. Concrete examples of the SQL queries run for a single dbt node (model, seed, snapshot) can be found in the unit tests.
 
Looking at `integration/common/tests/dbt/structured_logs/postgres/run/logs/successful_run_logs.jsonl` that represents the logs of the following dbt command (PG adapter)
```bash 
dbt --log-format json run --profiles-dir profiles
```
We can see for the node `model.jaffle_shop.orders` using this jq query:
```bash
jq -c 'select(.data.node_info.unique_id == "model.jaffle_shop.orders" and .info.name == "SQLQuery").data.sql' \
 "integration/common/tests/dbt/structured_logs/postgres/run/logs/successful_run_logs.jsonl"
```
The following SQL is executed (I've added comments and spaces for clarity)
```sql
--- SQL 1

BEGIN

--- SQL 2
/* {"app": "dbt", "dbt_version": "1.8.2", "profile_name": "jaffle_shop", "target_name": "dev", "node_id": "model.jaffle_shop.orders"} */

  create  table "postgres"."public"."orders__dbt_tmp"
    as(

with orders as (

    select * from "postgres"."public"."stg_orders"

),

payments as (

    select * from "postgres"."public"."stg_payments"

),

order_payments as (
    select
        order_id,
        sum(case when payment_method = 'credit_card' then amount else 0 end) as credit_card_amount,
        sum(case when payment_method = 'coupon' then amount else 0 end) as coupon_amount,
        sum(case when payment_method = 'bank_transfer' then amount else 0 end) as bank_transfer_amount,
        sum(case when payment_method = 'gift_card' then amount else 0 end) as gift_card_amount,
        sum(amount) as total_amount

    from payments

    group by order_id

),

final as (

    select
        orders.order_id,
        orders.customer_id,
        orders.order_date,
        orders.status,
        order_payments.credit_card_amount,
        order_payments.coupon_amount,
        order_payments.bank_transfer_amount,
        order_payments.gift_card_amount,
        order_payments.total_amount as amount

    from orders
    left join order_payments
        on orders.order_id = order_payments.order_id
)

select * from final
  );

-- SQL 3

/* {"app": "dbt", "dbt_version": "1.8.2", "profile_name": "jaffle_shop", "target_name": "dev", "node_id": "model.jaffle_shop.orders"} */

alter table "postgres"."public"."orders" rename to "orders__dbt_backup"

--  SQL 4

/* {"app": "dbt", "dbt_version": "1.8.2", "profile_name": "jaffle_shop", "target_name": "dev", "node_id": "model.jaffle_shop.orders"} */

alter table "postgres"."public"."orders__dbt_tmp" rename to "orders"

-- SQL 5

COMMIT

-- SQL 6

/* {"app": "dbt", "dbt_version": "1.8.2", "profile_name": "jaffle_shop", "target_name": "dev", "node_id": "model.jaffle_shop.orders"} */

drop table if exists "postgres"."public"."orders__dbt_backup" cascade
```

The OL events sent for those SQL queries can be found using the following jq command
```shell 
 jq -c '.[] | select(.job.facets.jobType.jobType == "SQL" and .run.facets.parent.job.name == "postgres.public.jaffle_shop.orders").job.facets.sql.query' \
 "integration/common/tests/dbt/structured_logs/postgres/run/results/successful_run_ol_events.json"
```
You would see two events one for START and another for COMPLETE.

>Regarding custom materializations that execute wildly different SQLs, I'm not sure - it sounds like that case makes processing more complex for the most common, simple case, without incremental materialization. I think it's valid use case...

Every dbt node is executed using a [materialization](https://github.com/dbt-labs/dbt-adapters/tree/main/dbt/include/global_project/macros/materializations) (table, view, seed, snapshot). These materializations don't only contain the rendered SQL of a given model. They do use [it](https://github.com/dbt-labs/dbt-adapters/blob/main/dbt/include/global_project/macros/materializations/models/table.sql#L31) (it's in the `sql` variable) but they have other queries on the side: 
* to create the temporary table that will hold the execution result of the rendered SQL
* to merge the temporary table with the target table
* ....etc 

>but it would be best if handling it would not cause worse experience for 99% of use cases - which IMO would be regular table/ephemeral/materialized_view materializations.

The common case already executes multiple SQL queries for the regular materializations. Users can override or create new [materializations](https://docs.getdbt.com/guides/create-new-materializations?step=2) as well.

The above is my understanding of dbt and how it executes its nodes. LMK if something is unclear. 

>So, maybe we should do
    if we're using non-incremental materializations, we should attach the sql execution data into regular START/COMPLETE events
    if we're using regular incremental materialization, we should intermittently emit RUNNING events for the same run - with the same run_id
    if we have some custom materialization... we should maybe have some custom dbt facet? I feel in any case dbt tries to keep to a model where one dbt node == one table, and single run for that IMO fits the best

IMO, the OL events sent should be agnostic of the materialization used for a given node.

Two solutions: 
1. we report on the SQL queries executed using the same run_id of the node via RUNNING events. 
2. we report on the SQL queries executed with their own OL START/COMPLETE events. Like demonstrated in this PR.

I'm in favor of 2 because:
1. It gives us more info on when a SQL query has been executed and when it finished. If we go with 1 consumers of OL will need to do some logic to compute the start/end times.
2. in the future we can add more info to those SQL OL events (dataplatform query_id,  session_id ...)

LMK.

Reply:
@MassyB that makes a lot of sense, thanks for describing it.

I would not be confident to turn this behavior on by default now, but what do you think about marking it as experimental, and describe it more in docs - website/docs/integrations/dbt.md file?

Given it's hidden behind `--consume-structured-logs`, and some ability to change the behavior in the future, I'd be good with accepting this PR.

Reply:
@JDarDagran @mobuchowski thanks again for the comments.

The [above](https://github.com/OpenLineage/OpenLineage/pull/3314#discussion_r1890476933) explains how dbt materializes its nodes (models, tests, snapshots). We see that multiple queries are executed for a given [materialization](https://docs.getdbt.com/docs/build/materializations). 

Let's discuss how we can model these SQL queries. The goal, in my opinion, is to maximize the observability capabilities of OL consumers.

Three solutions are discussed above (let me know if I forgot/misunderstood  one):

1. Wait for the [NodeFinished](https://github.com/dbt-labs/dbt-core/blob/95c090bed0b745c272a977d11fc248168e94bb25/core/dbt/events/types.py#L1516) event before sending all the SQL queries with the COMPLETE OL event.

```json

{"eventType": "START" ...} # on NodeStart dbt event

{"eventType": "COMPLETE", "job.facets.sql": ["query#1", "query#2", "query#N"] ...} # on NodeFinished dbt event

```

2. Send SQL queries as RUNNING events with the **same runId as the START event**. Each OL event would contain a facet for executed SQL queries. These events are sent when [SQLQuery](https://github.com/dbt-labs/dbt-adapters/blob/5741dc20dcf8c69aae77484383900c9d9fab133e/dbt/adapters/events/types.py#L180) dbt events are encountered.

```json

{"eventType": "START", "runId": "XXXX" ...} # on NodeStart dbt event

{"eventType": "RUNNING", "runId": "XXXX", "job.facets.sql": "query#1" ...} # on SQLQuery dbt event

{"eventType": "RUNNING", "runId": "XXXX", "job.facets.sql": "query#2" ...} # on SQLQuery dbt event

...

{"eventType": "RUNNING", "runId": "XXXX", "job.facets.sql": "query#n" ...} # on SQLQuery dbt event

{"eventType": "COMPLETE", "runId": "XXXX", ...} # on NodeFinished dbt event

```

3. Send SQL query OL events START/COMPLETE. They are sent on [SQLQuery](https://github.com/dbt-labs/dbt-adapters/blob/5741dc20dcf8c69aae77484383900c9d9fab133e/dbt/adapters/events/types.py#L180)/[SQLQueryStatus](https://github.com/dbt-labs/dbt-adapters/blob/5741dc20dcf8c69aae77484383900c9d9fab133e/dbt/adapters/events/types.py#L188)/[CatchableExceptionOnRun](https://github.com/dbt-labs/dbt-core/blob/95c090bed0b745c272a977d11fc248168e94bb25/core/dbt/events/types.py#L1775) dbt events. Each one has its own `runId`

```json

{"eventType": "START", "runId": "XXXX" ... } # on NodeStart dbt event

{"eventType": "START", "runId": "YYYY", "sql": "query#1" } # on SQLQuery dbt event

{"eventType": "COMPLETE", "runId": "YYYY", "sql": "query#1" ... } # on SQLQueryStatus dbt event

...

{"eventType": "START", "runId": "ZZZZ", "sql": "query#1" ... } # on SQLQuery dbt event

{"eventType": "COMPLETE", "runId": "ZZZZ", "sql": "query#n" ... } # on SQLQueryStatus dbt event

{"eventType": "COMPLETE", "runId": "XXXX" ... } # on NodeFinished dbt event

```

From the solutions listed above, I would go with number 3 because:

1. It allows consumers to have finer-grained observability on the SQL executed inside a dbt node.

2. It allows consumers to know as soon as possible when SQL is submitted for execution on the data platform, thanks to the OL START event.

3. It allows consumers to easily compute the elapsed time spent on a SQL query since both START/COMPLETE share the same `runId`.

4. It gives us the ability to add more facets dedicated to these SQL queries (query_id, session_id, etc.). These attributes would be adapter (Snowflake, Postgres, etc.) specific.

I wouldn't go with option 1 because:

1. It doesn't give us real-time monitoring. It waits for the node to finish before reporting on progress.

I wouldn't go with option 2 because:

1. It requires extra logic on the consumer side to get the time spent on each query.

2. It doesn't provide enough flexibility to report on the status of a single query. Failure would need to be reported on the node OL event. Success of `query#i` is deduced when we receive a RUNNING event of `query#i+1`.

3. Consumers need to keep an ordered list of the RUNNING events.

4. Facets that make sense for SQL queries would end up added to the node OL events.

In terms of scale, option 1 is not going to add a significant amount of events for consumers to process. From dbt's source code, the [materializations](https://github.com/dbt-labs/dbt-adapters/tree/main/dbt/include/global_project/macros/materializations) contain around 10 SQL queries.

What do you think? Which option do you prefer and why? Let me know if this is not the right place to discuss this.

I'm okay merging the PR since it's implementing option 3; however, I would like us to agree on a solution.

Thank you.

Reply:
@MassyB thanks for the writeup. I wasn't convinced to that solution initially, but after reading this I agree that it does seem to be the best option. Overcoming drawbacks of options 1 or 2 would require us to add extra facets, and that added complexity does not seem to make sense compared to added complexity of option 1.